docs/notes/router-annotations.adoc

```diff
@@ -0,0 +1,128 @@
+////
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License
+////
+
+= Router Message Annotations
+
+This document describes the implementation of the meta-data included
+in messages passed between router nodes (interior and edge).
+
+== Description
+
+Each message passed between routers includes meta-data used to
+communicate per-message state. This data is used by routers along the
+forwarding path for various reasons.  Currently this information
+includes:
+
+* The identifier for the ingress router (the router attached to the
+original sending endpoint).
+
+* The trace list: an ordered list of router identifiers. This list
+reflects the path the message has travelled in the router network. The
+first entry on the list is the ingress router *(Q: can we use this*
+*instead of a dedicated ingress entry field?)*.  The current router identifier
```

Comment:
> Q: can we use this instead of a dedicated ingress entry field?

Isn't reading a single field more efficient than decoding a list and reading the first item? (Regarding proton-c capabilities.) So if you sometimes want to only know the ingress router without knowing the full trace, separate field maybe can be more performant?

Reply:
It would be more efficient, yes.  But I don't believe there's any case where we don't need to decode the trace list - we need it to do the proper forwarding of the message.